wip-trampoline.md

```diff
@@ -0,0 +1,842 @@
+
+| Bits  | Name                 | Description                                        | Context | Link |
+|-------|----------------------|----------------------------------------------------|---------|------|
+| 50/51 | `trampoline_routing` | This node requires/supports trampoline routing     | IN9     |      |
+| 52/53 | `gossip_filters`     | This node requires/supports gossip message filters | IN      |      |
+
+Feature flags are assigned in pairs, but the mandatory value may not be applicable here.
+
+## Packet Structure
+
+### Trampoline Onion
+
+The trampoline onion is a fixed-size tlv packet with the following structure:
+
+1. type: 14 (`trampoline_onion_packet`)
+2. data:
+   * [`byte`:`version`]
+   * [`point`:`public_key`]
+   * [`400*byte`:`hop_payloads`]
+   * [`32*byte`:`hmac`]
+
+This results in a `466`-bytes fixed-size tlv where the value has exactly the same format as the `onion_packet` with
+a smaller `hop_payloads` (`400` bytes instead of `1300`).
+
+Trampoline `hop_payload`s may contain the following fields:
+
+1. tlvs: `trampoline_payload`
+2. types:
+    * type: 2 (`amt_to_forward`)
+    * data:
+        * [`tu64`:`amt_to_forward`]
+    * type: 4 (`outgoing_cltv_value`)
+    * data:
+        * [`tu32`:`outgoing_cltv_value`]
+    * type: 8 (`payment_data`)
+        * [`32*byte`:`payment_secret`]
+        * [`tu64`:`total_msat`]
+    * type: 10 (`outgoing_node_id`)
+    * data:
+        * [`point`:`outgoing_node_id`]
+    * type: 11 (`outgoing_node_features`)
+    * data:
+        * [`length*byte`:`outgoing_node_features`]
+
+### Paying via trampoline nodes
+
+A node that wants to rely on trampoline nodes to relay payments should use a `trampoline_onion_packet` in the
+`hop_payload` of the _last_ hop of a normal `onion_packet`:
+
+1. type: `onion_packet`
+2. data:
+   * [`byte`:`version`]
+   * [`point`:`public_key`]
+   * [`n1*byte`:`hop_payload`] (normal hop payload)
+   * [`n2*byte`:`hop_payload`] (normal hop payload)
+   * ...
+   * [`nn*byte`:`hop_payload`] (hop payload containing a `trampoline_onion_packet`)
+   * `filler`
+   * [`32*byte`:`hmac`]
+
+### Paying to non-trampoline nodes
+
+If the recipient of the payment doesn't support trampoline routing, the last trampoline node must convert the last hop
+to a standard onion payment. The `outgoing_node_features` field is used to detect that the next node doesn't support
+trampoline.
+
+Note that this reveals the identity of the recipient and the amount paid to the last trampoline node (but it doesn't
+reveal the identity of the payer). It also means that we'll need to add the invoice routing hints to the payload for
+the last trampoline node, otherwise it's likely that it won't be able to find a route to the recipient.
+
+Recipients should support `trampoline_routing` to properly preserve their anonymity.
+
+Question: should we even allow sending to non-trampoline-aware nodes? Maybe it's best to disallow it entirely.
```

Comment:
I really dislike that you can be paid via trampoline and without knowing it, your identity and the amount are revealed to some hop. Not only you don't know, but also you cannot opt out of this. This doesn't feel right to me.
Additionally, requiring the recipient to be trampoline aware should simplify the protocol.
If we hope for all nodes to eventually be trampoline-aware, this shouldn't be a big deal in the long term anyway.

So my preference would be to simply disallow it entirely. Or in other words, requiring recipients of trampoline payments to be trampoline-aware.

Reply:
I agree with @jtimon here.  This tends to put more load on each node, requiring each node to actually have a local routemap.  I have a number of ideas to compress routemaps as well.

Reply:
I agree too, I initially added that to make the upgrade smoother but I'd like to remove it before spec acceptance.
I already have another comment on this PR from December 5th where I stated that ;)

> I'm leaning more and more towards removing the ability to pay legacy nodes via this trampoline-to-legacy mechanism (unless we find a better way to do it than what I currently have).

It feels weird and it's bad for privacy. It will be removed from the proposal in the next iteration.